concurrency/MemberReceiver.java

```diff
@@ -0,0 +1,2351 @@
+    private final int opusChannels = 2;
+    private int frameSizeInSamplesPerChannel = (opusSampleRate * frameSizeInMillis) / 1000;
+    private int frameSizeInBytes = outputFrameSize * opusChannels * frameSizeInSamplesPerChannel;
+
+
+    private int dropPackets;
+
+    private boolean done = false;
+
+    private int myMemberNumber;
+    private static int memberNumber;
+    private static Object memberNumberLock = new Object();
+
+    /*
+     * Statistics
+     */
+    private String timeStarted;
+
+    private int packetsReceived = 0;
+    private int packetsDropped = 0;
+    private int comfortPayloadsReceived = 0;
+    private int comfortPayloadsSent = 0;
+
+    private long timeCurrentPacketReceived;
+    private long timePreviousPacketReceived;
+    private long totalTime = 0;
+    private long timeToProcessMediaPackets;
+    private int mediaPacketsReceived;
+    private int lastMediaPacketsReceived;
+    private int badPayloads;
+
+    private long previousRtpTimestamp = 0;
+
+    private boolean joinConfirmationReceived = false;
+
+    private static String joinConfirmationKey = "1";
+
+    private Cipher decryptCipher;
+
+    private String encryptionKey;
+    private String encryptionAlgorithm;
+
+    private SampleRateConverter inSampleRateConverter;
+
+    private DatagramChannel datagramChannel;
+    private SelectionKey selectionKey;
+    private RtcpReceiver rtcpReceiver;
+
+    private JitterManager jitterManager;
+
+    private ArrayList<MemberSender> forwardMemberList = new ArrayList<MemberSender>();
+
+    private boolean initializationDone = false;
+
+    private IChannel relayChannel = null;
+
+
+
+    public void setChannel(IChannel relayChannel)
+    {
+        this.relayChannel = relayChannel;
+    }
+
+    public MemberReceiver(ConferenceMember member, CallParticipant cp, DatagramChannel datagramChannel) throws IOException
+    {
+
+    this.member = member;
+    this.cp = cp;
+    this.datagramChannel = datagramChannel;
+
+    synchronized (memberNumberLock) {
+        myMemberNumber = memberNumber++;
+    }
+
+    encryptionKey = cp.getEncryptionKey();
+    encryptionAlgorithm = cp.getEncryptionAlgorithm();
+
+    if (encryptionKey != null) {
+        try {
+        if (encryptionKey.length() < 8) {
+            encryptionKey +=
+            String.valueOf(System.currentTimeMillis());
+        }
+
+        if (encryptionKey.length() > 8 &&
+            encryptionAlgorithm.equals("DES")) {
+
+            encryptionKey = encryptionKey.substring(0, 8);
+        }
+
+        byte[] keyBytes = encryptionKey.getBytes();
+        SecretKeySpec secretKey = new SecretKeySpec(keyBytes,
+            encryptionAlgorithm);
+
+            decryptCipher = Cipher.getInstance(encryptionAlgorithm);
+            decryptCipher.init(Cipher.DECRYPT_MODE, secretKey);
+
+        Logger.println("Call " + cp + " Voice data will be decrypted "
+            + "using " + encryptionAlgorithm);
+        } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)